src/core/server.cc

```diff
@@ -294,150 +293,354 @@ class HTTPServiceImpl {
   {
   }
 
-  tfs::net_http::RequestHandler Dispatch(
-      tfs::net_http::ServerRequestInterface* req);
+  static void Dispatch(evhtp_request_t* req, void* arg);
+
+  tensorflow::Status Start(uint16_t port, int thread_cnt) override;
+
+  tensorflow::Status Stop() override;
 
  private:
-  void Handle(tfs::net_http::ServerRequestInterface* req);
-  tfs::net_http::HTTPStatusCode Health(
-      tfs::net_http::ServerRequestInterface* req,
-      const std::string& health_uri);
-  tfs::net_http::HTTPStatusCode Profile(
-      tfs::net_http::ServerRequestInterface* req,
-      const std::string& profile_uri);
-  tfs::net_http::HTTPStatusCode Infer(
-      tfs::net_http::ServerRequestInterface* req, const std::string& infer_uri);
-  tfs::net_http::HTTPStatusCode Status(
-      tfs::net_http::ServerRequestInterface* req,
-      const std::string& status_uri);
+  void Handle(evhtp_request_t* req);
+  void Health(evhtp_request_t* req, const std::string& health_uri);
+  void Profile(evhtp_request_t* req, const std::string& profile_uri);
+  void Infer(evhtp_request_t* req, const std::string& infer_uri);
+  void Status(evhtp_request_t* req, const std::string& status_uri);
+  static std::string ToMethodString(htp_method method);
+
+  static void thread_init(evhtp_t* htp, evthr_t* thread, void* arg);
+  static void thread_exit(evhtp_t* htp, evthr_t* thread, void* arg);
+
+  // class object associated to evhtp thread, requests received are bounded
+  // with the thread that accepts it. Need to keep track of that and let the
+  // corresponding thread send back reply
+  class ThreadWiseService {
+   public:
+    class Request {
+     public:
+      Request(
+          evhtp_request_t* req, uint64_t id,
+          std::shared_ptr<HTTPInferRequestProvider> request_provider,
```

Comment:
Since you are immediately making a copy of these shared pointers you should pass them by const reference. That will avoid an unnecessary shared-pointer copy (I know that some recommend never using const& for shared_ptrs but I don't think that is correct advice).

Reply:
Depends on if you are capturing ownership or not.  Passing a reference to a shared_ptr is a little like passing a raw pointer since it could disappear out from under you.

But I agree, I pass references to shared_ptrs quite a bit, but only when I can guarantee the lifecycle of the object.